cmake/Subdirectories.cmake

```diff
@@ -191,6 +189,80 @@ function(add_registered_gems_to_external_subdirs output_gem_dirs gem_names)
     set(${output_gem_dirs} ${gem_dirs} PARENT_SCOPE)
 endfunction()
 
+#! Queries the list of gem names against the list of ALL registered external subdirectories
+#! in order to determine the paths corresponding to the gem names
+function(add_registered_gems_to_external_subdirs output_gem_dirs gem_names)
+    get_all_external_subdirectories(registered_external_subdirs)
+    query_gem_paths_from_external_subdirs(gem_dirs "${gem_names}" "${registered_external_subdirs}")
+    set(${output_gem_dirs} ${gem_dirs} PARENT_SCOPE)
+endfunction()
+
+#! Recurses "dependencies" array if the external subdirectory is a gem(contains a gem.json)
+#! for each subdirectory in use.
+#! This function looks up the each dependent gem path from the registered external subdirectory set
+#! That list of resolved gem paths then have this function invoked on it to perform the same behavior
+#! When every descendent gem referenced from the "dependencies" field of the current subdirectory is visited
+#! it is then appended to a list of output external subdirectories
+#! NOTE: This must be invoked after all the add_*_json_external_subdirectories function
+function(reorder_dependent_gems_with_cycle_detection _output_external_dirs subdirs_in_use registered_external_subdirs cycle_detection_set)
```

Comment:
Nit: I'm curious why this one variable has a leading `_` in its name.  Seems out of place with most of our CMake code.

Reply:
Yeah, this is needed because the caller to this function passed in a name of `output_external_dirs`, which when the double variable resolution occurs, resulted in just the variable name being output.